### PR TITLE
Add llm integration with Intel Gaudi in llama-index-llms-gaudi

### PR DESCRIPTION
**Pull request by jeanyu-habana**
# Description

This PR adds Intel Gaudi support to the list of llama_index llm integration. This enables users to run llm locally on Intel Gaudi using llama_index modules. 

Fixes # (issue)
N/A
## New Package?
None
Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [X] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [X] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [X] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

**Comment by logan-markewich** [comment]
@jeanyu-habana can you run `make lint` ? Many linting issues

File: llama-index-integrations/llms/llama-index-llms-gaudi/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+## Installation
+
+```bash
+pip install --upgrade-strategy eager optimum[habana]
+```
+
+## Usage
+
+```python
+from llama_index.llms.gaudi import GaudiLLM
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-gaudi/README.md:12]: Should we show how to use the LLM beyond importing it here?

File: llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/base.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+from llama_index.llms.gaudi.utils import initialize_model
+from llama_index.llms.huggingface import HuggingFaceLLM
+
+#DEFAULT_HUGGINGFACE_MODEL = "meta-llama/Llama-2-7b-chat-hf"
+DEFAULT_HUGGINGFACE_MODEL = "/home/ubuntu/jean/models/mistral"
+
+logger = logging.getLogger(__name__)
+
+
+class GaudiLLM(CustomLLM):
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/base.py:62]: This seems like mostly a copy of HuggingFaceLLM ? Is there any reason that a user can't directly use a HuggingFaceLLM with these models? If its just a difference in the `init`, should we just subclass `HuggingFaceLLM` to reduce code duplication?

File: llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/base.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+    StoppingCriteria,
+    StoppingCriteriaList,
+)
+from transformers import AutoTokenizer, LlamaTokenizer
+#gaudi
+from llama_index.llms.gaudi.utils import initialize_model
+from llama_index.llms.huggingface import HuggingFaceLLM
+
+#DEFAULT_HUGGINGFACE_MODEL = "meta-llama/Llama-2-7b-chat-hf"
+DEFAULT_HUGGINGFACE_MODEL = "/home/ubuntu/jean/models/mistral"
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/base.py:57]: Default model should not point to a local path

File: llama-index-integrations/llms/llama-index-llms-gaudi/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+packages = [{include = "llama_index/"}]
+readme = "README.md"
+version = "0.1.0"
+
+[tool.poetry.dependencies]
+python = ">=3.8.1,<4.0"
+huggingface-hub = "^0.23.0"
+torch = "^2.1.2"
+text-generation = "^0.7.0"
+llama-index-core = "^0.11.0"
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-gaudi/pyproject.toml:40]: `llama-index-llms-huggingface` is imported but not listed as a dependency

File: llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/utils.py [hunk 5]
```diff
@@ -0,0 +1,788 @@
...
+    write_checkpoints_json,
+)
+from optimum.habana.utils import (
+    check_habana_frameworks_version,
+    check_optimum_habana_min_version,
+    get_habana_frameworks_version,
+    set_seed,
+)
+
+def setup_parser(parser):
```
[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/utils.py:47]: Just for the example I guess?

File: llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/utils.py [hunk 6]
```diff
@@ -0,0 +1,788 @@
...
+import shutil
+import tempfile
+import time
+from pathlib import Path
+
+import torch
+from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer
+from transformers.utils import check_min_version
+
+from optimum.habana.checkpoint_utils import (
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/utils.py:33]: Is this import covered by the current dependencies?

**Comment by jeanyu-habana** [comment]
> @jeanyu-habana can you run `make lint` ? Many linting issues

Ran "make lint" and fixed errors.

File: llama-index-integrations/llms/llama-index-llms-gaudi/README.md (hunk 1 above)
Comment (jeanyu-habana) [reply llama-index-integrations/llms/llama-index-llms-gaudi/README.md:16]: Updated this README to include code example beyond imports.

File: llama-index-integrations/llms/llama-index-llms-gaudi/pyproject.toml (hunk 4 above)
Comment (jeanyu-habana) [reply llama-index-integrations/llms/llama-index-llms-gaudi/pyproject.toml:40]: Updated the pyproject.toml file to include llama-index-llms-huggingface

File: llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/base.py (hunk 2 above)
Comment (jeanyu-habana) [reply llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/base.py]: I agree with you. Updated base.py to use HuggingFaceLLM as base class.

File: llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/base.py (hunk 3 above)
Comment (jeanyu-habana) [reply llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/base.py]: Replaced the local path.

File: llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeanyu-habana) [reply llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/utils.py]: Yes, the parser is for example and it should stay with the example. Fixed utils.py to remove the CLI parser. Now example has the CLI parser.

File: llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/utils.py (hunk 6 above)
Comment (jeanyu-habana) [reply llama-index-integrations/llms/llama-index-llms-gaudi/llama_index/llms/gaudi/utils.py:31]: Updated pyproject.toml to include optimum habana dependency.

**Comment by jeanyu-habana** [comment]
> > @jeanyu-habana can you run `make lint` ? Many linting issues
> 
> Ran "make lint" and fixed errors.

Done.

**Comment by jeanyu-habana** [comment]
@logan-markewich Would you be able take another look at the changes made and let me know if previous comments can be resolved? I'd appreciate it.

**Review by logan-markewich — APPROVED** [review]

**Comment by logan-markewich** [comment]
We added some checks for unit tests 😅 Would it be possible to add some?

Could be similar to what I did here
https://github.com/run-llama/llama_index/pull/16413

**Comment by logan-markewich** [comment]
Actually, seems you escape the check for now

**Pull request merged**